returnn/tf/layers/basic.py

```diff
@@ -854,44 +858,87 @@ class SliceNdLayer(_ConcatInputLayer):
 
   def __init__(self, start, size, min_size=None, **kwargs):
     """
-    :param LayerBase start:
+    :param LayerBase start: (B,...)
     :param int|None size: if None, it uses the max possible size, and it becomes a dynamic axis
-    :param int|None min_size: if size is None, but we want to have a min-size, set this
+    :param int|None min_size: if size is None, but we want to have a min-size
     """
     super(SliceNdLayer, self).__init__(**kwargs)
-    from returnn.tf.util.basic import slice_nd, where_bc, expand_multiple_dims, DimensionTag
+    from returnn.tf.util.basic import where_bc, expand_multiple_dims
     x = self.input_data.copy_as_batch_major()
-    assert x.time_dim_axis == 1, "currently only time-axis==1 supported"
-    seq_lens = x.get_sequence_lengths() if x.is_time_axis_dynamic() else None
+    seq_lens = x.get_sequence_lengths() if x.is_time_axis_dynamic() else None  # (B,) or None
     self.start = start
-    assert start.output.have_batch_axis() and start.output.batch_shape == (None,)
-    start = start.output.get_placeholder_as_batch_major()
+    start_data = start.output.copy_as_batch_major()  # e.g. (B,) or (B,T)
+    start_t = start_data.placeholder
     if size is None:
       if seq_lens is None:
-        size = tf.maximum(tf.reduce_max(x.batch_shape[1] - start), 0)
+        size = tf.maximum(tf.reduce_max(x.batch_shape[1] - start_t), 0)  # scalar
       else:
-        size = tf.maximum(tf.reduce_max(seq_lens - start), 0)
+        # make seq_lens compatible with start_t
+        seq_lens = expand_multiple_dims(  # e.g. (B,) or (B,1)
+          x=seq_lens,
+          axes=[-1] * (len(start_t.shape) - len(seq_lens.shape)))
+        size = tf.maximum(tf.reduce_max(seq_lens - start_t), 0)  # scalar
       if min_size is not None:
         size = tf.maximum(size, min_size)
-    self.size = size
-    start = tf.expand_dims(start, axis=1)  # (B, T)
-    slices = slice_nd(x.placeholder, start=tf.cast(start, tf.int32), size=size)  # (B,size, ...)
+    # for each start index in start_data, we want to gather a slice
+    # therefore, the output's first axes are the same as the ones from start_data
+    # and the next axis will therefore be the slice axis
+    slice_tag = self.output.dim_tags[start_data.batch_ndim]
+    assert slice_tag.description.startswith("sliced-time:")
+    if not isinstance(size, int):
+      # in this case, size is not known before runtime and becomes dynamic and we need to set dyn_size
+      if len(seq_lens.shape) == 1:
```

Comment:
I just notice, you do not really cover the case where `seq_lens is None` here, right?

Reply:
Yes, you are right. I am going to fix this

Reply:
Why do you have this check anyway? Isn't the following code always valid also for `len(seq_lens.shape) != 1`?

Reply:
I am not sure if `tf.reduce_max` works if `reduce_axes` is `range(1,1)` and therefore empty. I have to check this.

Reply:
Why `reduce_max`? There is no `reduce_max` in the following code (`dyn_size = tf.maximum(seq_lens - start_t, 0)`).

Reply:
> There is no `reduce_max` in the following code

But there is a `reduce_max` in the `else` case and I wasn't sure if this would work for `len(seq_lens.shape) != 1`. But I just checked and it works. But we leave out the axis reduction anyway so we don't need the condition.

Reply:
Yes sorry, that is what I meant with the following code, i.e. only the true-branch, not the `else` branch.

Reply:
Oh and I just understood what you meant. Yes, I could have first done the `dyn_size = tf.maximum(seq_lens - start_t, 0)` without the condition and then, in the case of `len(seq_lens.shape) != 1`, I could have done the reduction.

Reply:
No, I meant above ("Isn't the following code always valid also for `len(seq_lens.shape) != 1`?) whether the true branch code is correct both both cases. Which implies that the false branch code is wrong. And it implies that the check is not needed.